File: 天梯赛/L2-003. 月饼.cpp
#include<iostream>
#include<vector>
#include<cmath>
#include<algorithm>
using namespace std;
const int maxn=510;
const int inf=0x3f3f3f3f;
struct node{
	double save;
	double sale;
	double unite;
};
bool cmp(node a,node  b){
	return a.unite>b.unite;
}
int main(){
	int n,d,in=0;
	double maxsum=0;
	cin>>n>>d;
	vector<node> v(n);
	for(int i=0;i<n;++i)cin>>v[i].save;
	for(int i=0;i<n;++i){cin>>v[i].sale;v[i].unite=v[i].sale/v[i].save;}
	sort(v.begin(),v.end(),cmp);
	while(d>0&&in<v.size()){
		if(v[in].save<=d){
			d-=v[in].save;
			maxsum+=v[in].sale;
			++in;
		}else{
			maxsum+=v[in].unite*d;
			d=0;
		}
	}
	printf("%.2lf",maxsum);
	return 0;
}

File: 天梯赛/java/L2-003. 月饼.java
import java.util.Arrays;
import java.util.Scanner;
import java.util.Vector;

public class Main {
	public static final int maxn=510;
	public static final int inf=0x3f3f3f3f;
	public static void main(String[] args) {
		Scanner in=new Scanner(System.in);
		int n=in.nextInt(),d=in.nextInt();
		node v[]=new node [n];
		for(int i=0;i<n;++i) {
			v[i]=new node();
			v[i].save=in.nextDouble();
		}
		for(int i=0;i<n;++i) {
			v[i].sale=in.nextDouble();
			v[i].unite=v[i].sale/v[i].save;
		}
		in.close();
		Arrays.sort(v);
		int i=0;
		double maxsum=0;
		while(d>0&&i<v.length) {
			if(v[i].save<=d) {
				d-=v[i].save;
				maxsum+=v[i].sale;
				++i;
			}else {
				maxsum+=v[i].unite*d;
				d=0;
			}
		}
		System.out.printf("%.2f",maxsum);
	}
}

class node implements Comparable<node>{
	double save;
	double sale;
	double unite;
	
	@Override
	public int compareTo(node a) {
		if(unite>a.unite)return -1;
		return 1;
	}
}